torch/sparse/semi_structured.py

```diff
@@ -364,17 +364,20 @@ def __torch_dispatch__(cls, func, types, args, kwargs) -> Any:
         raise NotImplementedError(error_string)
 
     def to_dense(self):
+        if not self._FORCE_CUTLASS:
```

Comment:
I think we want to check against what _FORCE_CUTLASS was set to when the tensor was created, not the current state. 

I'm thinking something like `self._created_with_cutlass` which gets set when the subclass is created.

Reply:
Thanks for the correction.  While we're at it:  How about adding `self.meta_cutlass` member variable too, and then avoid forcing CUTLASS backend to keep both sparse matrix and metadata in the same tensor?  This way, we could get rid of compressed size calculations when initializing for CUTLASS backed, and in general this initialization could be made simpler.  I also somewhat suspect that issues like [this](https://github.com/pytorch/pytorch/issues/105540) may be resolved this way.

Reply:
Here is a patch, to be applied on top of current state of this PR, implementing what I'm proposing - it adds both `created_with_cutlass` and `meta_tensor_cutlass` member variables:

<details>
 <summary>Patch</summary>

```diff
diff --git a/torch/sparse/semi_structured.py b/torch/sparse/semi_structured.py
index b9ab353b6c1..66e44f98abc 100644
--- a/torch/sparse/semi_structured.py
+++ b/torch/sparse/semi_structured.py
@@ -10,14 +10,14 @@ __all__ = [
 ]
 
 _SEMI_STRUCTURED_SPARSE_CONFIG = namedtuple(
-    "_SEMI_STRUCTURED_SPARSE_CONFIG", "compression_factor min_rows min_cols"
+    "_SEMI_STRUCTURED_SPARSE_CONFIG", "min_rows min_cols"
 )
 _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG = {
-    torch.int8: _SEMI_STRUCTURED_SPARSE_CONFIG(10, 32, 128),
-    torch.float16: _SEMI_STRUCTURED_SPARSE_CONFIG(9, 32, 64),
-    torch.bfloat16: _SEMI_STRUCTURED_SPARSE_CONFIG(9, 32, 64),
+    torch.int8: _SEMI_STRUCTURED_SPARSE_CONFIG(32, 128),
+    torch.float16: _SEMI_STRUCTURED_SPARSE_CONFIG(32, 64),
+    torch.bfloat16: _SEMI_STRUCTURED_SPARSE_CONFIG(32, 64),
     # TODO enable float32 support when adding cuSPARSELt as a backend
-    # torch.float32: _SEMI_STRUCTURED_SPARSE_CONFIG(9, 32, 32)
+    # torch.float32: _SEMI_STRUCTURED_SPARSE_CONFIG(32, 32)
 }
 
 
@@ -32,17 +32,18 @@ class SparseSemiStructuredTensor(torch.Tensor):
     We also support 1:2 sparsity for float32 dtype.
 
     This subclass stores the dense tensor in a compressed form by only storing the specified elements and corresponding metadata.
-    These two are stored next to each other in one contiguous tensor.
 
-    We choose to store the specified elements and the metadata in a single tensor for compatibilty with cuSPARSELt,
-    which expects the data to be stored in this format.
+    The subclass supports two backend, either CUTLASS or cuSPASRELt.
+
+    The cuSPARSELt backend expects the specified elements and the metadata to be stored in a single tensor:
 
     compressed tensor = [ specified elements of original tensor | metadata ]
 
     For an original tensor of size (m, k) we expect the first m * k // 2 elements to be the kept elements
     The rest of the tensor is metadata.
 
-    The subclass supports two backends, either CUTLASS or cuSPASRELt.
+    For CUTLASS backend, the compressed tensor contains specified element of original tensor only,
+    while metadata is kept in separate tensor.
 
     When _FORCE_CUTLASS is set, or when cuSPARSELt is not available, this subclass calls into _sparse_semi_structured_linear
     and sparse_semi_structured_from_dense for conversion to the compressed format.
@@ -62,6 +63,8 @@ class SparseSemiStructuredTensor(torch.Tensor):
         original_shape: Optional[torch.Size] = None,
         compressed_tensor: Optional[torch.Tensor] = None,
         transposed: bool = False,
+        created_with_cutlass: bool = False,
+        meta_tensor_cutlass: Optional[torch.Tensor] = None,
     ):
         """
         Create a new instance of the class.
@@ -72,8 +75,11 @@ class SparseSemiStructuredTensor(torch.Tensor):
         Args:
             original_tensor: The original dense tensor, or None, if we have already compressed the tensor.
             original_shape: The shape of the original dense tensor
-            compressed_tensor: A flattened tensor to store the specified elements and metadata.
+            compressed_tensor: A flattened tensor to store the specified elements and metadata for cuSPARSELt backend,
+            or tensor with elements only for CUTLASS backend.
             transposed: Whether the tensor is transposed or not.
+            created_with_cutlass: Whether the tensor created for CUTLASS backend or not.
+            meta_tensor_cutlass: Unused for cuSPARSELt backend, and tensor to store metadata for CUTLASS backend.
 
         Returns:
             torch.Tensor: A torch.Tensor wrapper subclass.
@@ -126,14 +132,19 @@ class SparseSemiStructuredTensor(torch.Tensor):
         original_shape: Optional[torch.Size] = None,
         compressed_tensor: Optional[torch.Tensor] = None,
         transposed: bool = False,
+        created_with_cutlass: bool = False,
+        meta_tensor_cutlass: Optional[torch.Tensor] = None,
     ) -> None:
         """SparseSemiStructuredTensor constructor.
 
         Args:
             original_tensor: The original dense tensor, or None, if we have already compressed the tensor.
             original_shape: The shape of the original dense tensor
-            compressed_tensor: A flattened tensor to store the specified elements and metadata.
+            compressed_tensor: A flattened tensor to store the specified elements and metadata for cuSPARSELt backend,
+            or tensor with elements only for CUTLASS backend.
             transposed: Whether the tensor is transposed or not.
+            created_with_cutlass: Whether the tensor created for CUTLASS backend or not.
+            meta_tensor_cutlass: Unused for cuSPARSELt backend, and tensor to store metadata for CUTLASS backend.
 
         Returns:
             None
@@ -179,42 +190,28 @@ class SparseSemiStructuredTensor(torch.Tensor):
                 # TODO in the future we can add in padding to support dimensions that aren't perfect multiples
                 raise RuntimeError(
                     f"Error original_tensor.shape {original_tensor.shape} is not supported! "
-                    "Both dimensions must be larger or equal than and a multiple of ({min_rows}, {min_cols})"
+                    f"Both dimensions must be larger or equal than and a multiple of ({min_rows}, {min_cols})"
                 )
 
             if self._FORCE_CUTLASS:
-                # This code calculates the size of the compressed tensor.
-                # compression factor is different based on dtype it's given by the formula below for 2:4 sparsity:
-                # compression_factor = 1/2 + 1/bitwidth(dtype)
-                original_size = original_tensor.nelement()
-                compression_factor = _DTYPE_TO_SEMI_STRUCTURED_SPARSE_CONFIG[
-                    original_tensor.dtype
-                ].compression_factor
-                compressed_size = original_size * compression_factor // 16
-
-                compressed_tensor = torch.empty(
-                    (compressed_size,),
-                    dtype=original_tensor.dtype,
-                    device=original_tensor.device,
-                )
-
                 from torch.sparse._semi_structured_conversions import (
                     sparse_semi_structured_from_dense_cutlass,
                 )
 
-                sparse, meta = sparse_semi_structured_from_dense_cutlass(original_tensor)
-                compressed_tensor[: m * n // 2] = sparse.view(-1)
-                compressed_tensor[m * n // 2 :] = meta.view(original_tensor.dtype).view(
-                    -1
-                )
+                compressed_tensor, meta_tensor_cutlass = sparse_semi_structured_from_dense_cutlass(original_tensor)
+                created_with_cutlass = True
             else:
                 # use cuSPARSELt
                 compressed_tensor = torch._cslt_compress(original_tensor)
+                created_with_cutlass = False
+                meta_tensor_cutlass = None
 
         # set values
         self.original_tensor = None
         self.compressed_tensor = compressed_tensor
         self.transposed = transposed
+        self.created_with_cutlass = created_with_cutlass
+        self.meta_tensor_cutlass = meta_tensor_cutlass
 
     def __repr__(self) -> str:  # type: ignore[override]
         """Return string representation of SparseSemiStructuredTensor
@@ -260,6 +257,8 @@ class SparseSemiStructuredTensor(torch.Tensor):
                 original_shape=args[0].shape,
                 compressed_tensor=args[0].compressed_tensor,
                 transposed=args[0].transposed,
+                created_with_cutlass=args[0].created_with_cutlass,
+                meta_tensor_cutlass=args[0].meta_tensor_cutlass,
             )
 
         # Because we cannot go from the compressed representation back to the dense representation currently,
@@ -271,6 +270,8 @@ class SparseSemiStructuredTensor(torch.Tensor):
                 original_shape=args[0].shape,
                 compressed_tensor=args[0].compressed_tensor,
                 transposed=not args[0].transposed,
+                created_with_cutlass=args[0].created_with_cutlass,
+                meta_tensor_cutlass=args[0].meta_tensor_cutlass,
             )
 
         # handle addmm
@@ -341,21 +342,25 @@ class SparseSemiStructuredTensor(torch.Tensor):
 
         # handle values
         if func is torch.ops.aten.values.default:
-            m, k = args[0].shape
-            num_kept_elements = m * k // 2
-            return args[0].compressed_tensor[:num_kept_elements].view(m, k // 2)
+            if args[0].created_with_cutlass:
+                return args[0].compressed_tensor
+            else:
+                m, k = args[0].shape
+                num_kept_elements = m * k // 2
+                return args[0].compressed_tensor[:num_kept_elements].view(m, k // 2)
 
         # handle indices
         if func is torch.ops.aten.indices.default:
-            m, k = args[0].shape
-            num_kept_elements = m * k // 2
-            metadata = args[0].compressed_tensor[num_kept_elements:].view(m, -1)
-
-            # the metadata is expected to be in different datatypes for fp16/int8 respectively for CUTLASS.
-            indices_dtype = SparseSemiStructuredTensor.__get_indices_dtype(
-                args[0].dtype
-            )
-            return metadata.view(indices_dtype)
+            if args[0].created_with_cutlass:
+                return args[0].meta_tensor_cutlass
+            else:
+                m, k = args[0].shape
+                num_kept_elements = m * k // 2
+                metadata = args[0].compressed_tensor[num_kept_elements:].view(m, -1)
+                indices_dtype = SparseSemiStructuredTensor.__get_indices_dtype(
+                    args[0].dtype
+                )
+                return metadata.view(indices_dtype)
 
         error_string = "\n".join(
             [f"func {func} with args: "]
@@ -363,23 +368,21 @@ class SparseSemiStructuredTensor(torch.Tensor):
         )
         raise NotImplementedError(error_string)
 
+
     def to_dense(self):
-        if not self._FORCE_CUTLASS:
+        if not self.created_with_cutlass:
             raise RuntimeError("Converting to dense is not yet supported by cuSPARSELt backend!")
 
         if self.compressed_tensor is None:
             raise RuntimeError("Compressed tensor is not set, cannot convert to dense!")
 
-        m, n = self.shape
-        indices_dtype = SparseSemiStructuredTensor.__get_indices_dtype(self.dtype)
-
         from torch.sparse._semi_structured_conversions import (
             sparse_semi_structured_to_dense_cutlass,
         )
 
         return sparse_semi_structured_to_dense_cutlass(
-            self.compressed_tensor[: m * n // 2].view(m, -1),
-            self.compressed_tensor[m * n // 2 :].view(indices_dtype).view(m, -1),
+            self.compressed_tensor,
+            self.meta_tensor_cutlass,
         )
 
```
</details>

All the tests from `test_sparse_semi_structured.py`  pass with this patch, except for `test_values` test.  More precisely, whenever `values()` called, following error gets reported:
```
RuntimeError: Attempted to make a tensor into a differentiable view, but the tensor already had autograd metadata associated with it.  If you are using a __torch_dispatch__ mode, the most common cause for this problem is that you used torch.overrides.enable_reentrant_dispatch() improperly; tensors created within the extent of reentrant dispatch MUST NOT be directly returned from __torch_dispatch__; instead, they must be wrapped into fresh tensors that serve as the output.  If you are not using wrappers, you probably don't need reentrant dispatch.  If this doesn't seem applicable, please file a bug to PyTorch.
```
On the other side, calling `indices()` works fine.  I'm not sure what is the problem here, but on the other side I'm wondering do we need to override dispatch for `values()` and `indices()` at all: these seem to be intended for sparse COO tensors only, and for SparseSemiStructuredTensor we could keep them as internal methods of the class.

On an unrelated note: Do the CI at the moment builds with cuSPARSELt?

Reply:
As far as I know CI currently does not build with cuSPARSElt. cc @malfet

Reply:
cc @alexsamardzic 

I think it would be better to either store the CUTLASS values separately in the subclass than reusing `compressed_tensor`. i.e. -> something like cusparselt_compressed, cutlass_values, cutlass_metadata. Or even two different tensor subclasses, one for CUTLASS and one for cuSPARSELt. I think reusing compressed_tensor is kind of confusing. I agree that if we can't reuse the same code for reconstructing the dense matrix then separating cuSPASRELt from CUTLASS would make the code cleaner. 

I think implementing values/indicies as internal methods is fine, but it is nicer to do this with dispatching. 
Can you try `self.compressed_tensor = torch.clone(compressed_tensor)`? I think that may fix your issue.

Reply:
Changed according to your suggestion.  Now there is no need for `created_with_cutlass`, as `compressed_tensor_cusparselt` being `None` provides this information.  This is now ready for review/merge.

The problem with the `test_values` is still there on my machine, using `torch.clone()` doesn't seem to help.  On the other side, let's see if CI passes, maybe it's something about my setup only.

Reply:
Ok, CI still fails.  Fixed it with `detach()`, albeit `torch.clone()` also seems to work now.  I hope `detach()` is better, as it seems to keep the same underlying storage (the `data_ptr()` printed after `detach()` is the same as for original tensor, while this is not the case after `clone()`).  I still don't get what was the problem here, in any case this PR is fully ready for review/merge now.